fd.go

```diff
@@ -123,47 +134,63 @@ func (fi *fileDescriptor) Flush() error {
 
 // flushUp syncs the file and adds it to the dagservice
 // it *must* be called with the File's lock taken
-// TODO: What is `fullsync`? Propagate the changes upward
-// to the root flushing every node in the path (the "up"
-// part of `flushUp`).
+// If `fullSync` is set the changes are propagated upwards
+// (the `Up` part of `flushUp`).
 func (fi *fileDescriptor) flushUp(fullSync bool) error {
-	nd, err := fi.mod.GetNode()
-	if err != nil {
-		return err
-	}
+	var nd ipld.Node
+	switch fi.state {
+	case stateCreated, stateDirty:
+		var err error
+		nd, err = fi.mod.GetNode()
+		if err != nil {
+			return err
+		}
+		err = fi.inode.dagService.Add(context.TODO(), nd)
+		if err != nil {
+			return err
+		}
+
+		// Always update the file descriptor's inode with the created/modified node.
+		fi.inode.nodeLock.Lock()
+		fi.inode.node = nd
+		fi.inode.nodeLock.Unlock()
+
+		if !fullSync {
+			return nil
+		}
 
-	err = fi.inode.dagService.Add(context.TODO(), nd)
-	if err != nil {
-		return err
+		// Bubble up the update's to the parent, only if fullSync is set to true.
+		fi.inode.nodeLock.Lock()
+		nd = fi.inode.node
+		parent := fi.inode.parent
+		name := fi.inode.name
+		fi.inode.nodeLock.Unlock()
+
+		if err := parent.updateChildEntry(child{name, nd}); err != nil {
+			return err
+		}
+		fi.state = stateFlushed
+		return nil
+	case stateFlushed:
+		return nil
+	default:
+		panic("invalid state")
 	}
-	// TODO: Very similar logic to the update process in
```

Comment:
Please keep this comment, looking at the code again this is really *very* similar to the `localUpdate` logic, we'll need to find a way to abstract that (not in this PR tough).

Reply:
Will do.